File: apps/app-frontend/src/helpers/jre.js
/**
 * All theseus API calls return serialized values (both return values and errors);
 * So, for example, addDefaultInstance creates a blank instance object, where the Rust struct is serialized,
 *  and deserialized into a usable JS object.
 */
import { invoke } from '@tauri-apps/api/core'

/*

JavaVersion {
    path: Path
    version: String
}

*/

export async function get_java_versions() {
	return await invoke('plugin:jre|get_java_versions')
}

export async function set_java_version(javaVersion) {
	return await invoke('plugin:jre|set_java_version', { javaVersion })
}

// Finds all the installation of Java 7, if it exists
// Returns [JavaVersion]
export async function find_filtered_jres(version) {
	return await invoke('plugin:jre|jre_find_filtered_jres', { version })
}

// Gets java version from a specific path by trying to run 'java -version' on it.
// This also validates it, as it returns null if no valid java version is found at the path
export async function get_jre(path) {
	return await invoke('plugin:jre|jre_get_jre', { path })
}

// Tests JRE version by running 'java -version' on it.
// Returns true if the version is valid, and matches given (after extraction)
export async function test_jre(path, majorVersion) {
	return await invoke('plugin:jre|jre_test_jre', { path, majorVersion })
}

// Automatically installs specified java version
export async function auto_install_java(javaVersion) {
	return await invoke('plugin:jre|jre_auto_install_java', { javaVersion })
}

// Get max memory in KiB
export async function get_max_memory() {
	return await invoke('plugin:jre|jre_get_max_memory')
}

/**
 * Which GPU Windows hands a given Java runtime.
 *
 * `preference` mirrors the values the OS stores; which physical adapter
 * "high_performance" resolves to is the driver's decision, not ours.
 *
 * @typedef {'auto' | 'power_saving' | 'high_performance'} GpuPreference
 * @typedef {{ name: string, likely_discrete: boolean }} GpuAdapter
 * @typedef {{
 *   supported: boolean,
 *   unsupported_reason: string | null,
 *   java_path: string,
 *   preference: GpuPreference,
 *   adapters: GpuAdapter[],
 * }} JavaGpuStatus
 */

/** @returns {Promise<JavaGpuStatus>} */
export async function get_gpu_status(path) {
	return await invoke('plugin:jre|jre_get_gpu_status', { path })
}

/** @returns {Promise<JavaGpuStatus>} */
export async function set_gpu_preference(path, preference) {
	return await invoke('plugin:jre|jre_set_gpu_preference', { path, preference })
}

/**
 * A Java runtime the launcher downloaded, as it sits on disk.
 *
 * @typedef {{
 *   path: string,
 *   java_path: string,
 *   name: string,
 *   major_version: number | null,
 *   size_bytes: number,
 *   in_use: boolean,
 * }} InstalledRuntime
 */

/** @returns {Promise<InstalledRuntime[]>} */
export async function list_installed_runtimes() {
	return await invoke('plugin:jre|jre_list_installed_runtimes')
}

export async function remove_installed_runtime(path) {
	return await invoke('plugin:jre|jre_remove_installed_runtime', { path })
}

/** @returns {Promise<number>} bytes reclaimed */
export async function remove_unused_runtimes() {
	return await invoke('plugin:jre|jre_remove_unused_runtimes')
}

File: apps/app/src/api/jre.rs
use crate::api::Result;
use dashmap::DashMap;
use std::path::PathBuf;
use tauri::plugin::TauriPlugin;
use theseus::prelude::JavaVersion;
use theseus::prelude::*;

pub fn init<R: tauri::Runtime>() -> TauriPlugin<R> {
    tauri::plugin::Builder::new("jre")
        .invoke_handler(tauri::generate_handler![
            get_java_versions,
            set_java_version,
            jre_find_filtered_jres,
            jre_get_jre,
            jre_test_jre,
            jre_auto_install_java,
            jre_get_max_memory,
            jre_get_gpu_status,
            jre_set_gpu_preference,
            jre_list_installed_runtimes,
            jre_remove_installed_runtime,
            jre_remove_unused_runtimes,
        ])
        .build()
}

/// Every Java runtime the launcher has downloaded, with its size on disk.
#[tauri::command]
pub async fn jre_list_installed_runtimes() -> Result<Vec<jre::InstalledRuntime>>
{
    Ok(jre::list_installed_runtimes().await?)
}

#[tauri::command]
pub async fn jre_remove_installed_runtime(path: PathBuf) -> Result<()> {
    jre::remove_installed_runtime(path).await?;
    Ok(())
}

/// Removes every runtime no Java setting points at. Returns bytes reclaimed.
#[tauri::command]
pub async fn jre_remove_unused_runtimes() -> Result<u64> {
    Ok(jre::remove_unused_runtimes().await?)
}

/// Which GPU Windows will hand the runtime at `path`, and what is available.
#[tauri::command]
pub async fn jre_get_gpu_status(path: PathBuf) -> Result<gpu::JavaGpuStatus> {
    Ok(gpu::status(&path)?)
}

#[tauri::command]
pub async fn jre_set_gpu_preference(
    path: PathBuf,
    preference: gpu::GpuPreference,
) -> Result<gpu::JavaGpuStatus> {
    gpu::set_preference(&path, preference)?;
    Ok(gpu::status(&path)?)
}

#[tauri::command]
pub async fn get_java_versions() -> Result<DashMap<u32, JavaVersion>> {
    Ok(jre::get_java_versions().await?)
}

#[tauri::command]
pub async fn set_java_version(java_version: JavaVersion) -> Result<()> {
    jre::set_java_version(java_version).await?;
    Ok(())
}

// Finds the installation of Java 8, if it exists
#[tauri::command]
pub async fn jre_find_filtered_jres(
    version: Option<u32>,
) -> Result<Vec<JavaVersion>> {
    Ok(jre::find_filtered_jres(version).await?)
}

// Validates JRE at a given path
// Returns None if the path is not a valid JRE
#[tauri::command]
pub async fn jre_get_jre(path: PathBuf) -> Result<JavaVersion> {
    Ok(jre::check_jre(path).await?)
}

// Tests JRE of a certain version
#[tauri::command]
pub async fn jre_test_jre(path: PathBuf, major_version: u32) -> Result<bool> {
    Ok(jre::test_jre(path, major_version).await?)
}

// Auto installs java for the given java version
#[tauri::command]
pub async fn jre_auto_install_java(java_version: u32) -> Result<PathBuf> {
    Ok(jre::auto_install_java(java_version).await?)
}

// Gets the maximum memory a system has available.
#[tauri::command]
pub async fn jre_get_max_memory() -> Result<u64> {
    Ok(jre::get_max_memory().await?)
}
